### PR TITLE
feat/AB#105413_EIS-file-button-download-in-widgets-using-context

### PR DESCRIPTION
**Pull request by unai-reliefapp**
# Description

feat: refactor html parser logic for replace records fields in order to avoid the toReadableObject property save in memory on each field 
feat: add context values for editor and summary card html values. Before the parser, we extract any additional needed fields and values in order to be available for the parser, plus prepare the given html to be ready for replace value

## Useful links

- [Link to ticket](https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/105413)

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [X] Improvement (refactor or addition to existing functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

- [ ] Test A
- [ ] Test B

## Screenshots

Please include screenshots of this change. If this issue is only back-end related, and does not involve any visual change of the platform, you can skip this part.

# Checklist:

( * == Mandatory ) 

- [X] * I have set myself as assignee of the pull request
- [X] * My code follows the style guidelines of this project
- [X] * Linting does not generate new warnings
- [X] * I have performed a self-review of my own code
- [X] * I have put the ticket for review, adding the oort-frontend team to the list of reviewers
- [X] * I have commented my code, particularly in hard-to-understand areas
- [X] * I have put JSDoc comment in all required places
- [X] * My changes generate no new warnings
- [X] * I have included screenshots describing my changes if relevant
- [X] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# More explanation
https://www.loom.com/share/05a716d61b9744faaf51fb304c21d1e5?sid=f87cf896-582a-4f76-93ae-8ceed801b145

### CODE REVIEW

**Comment by unai-reliefapp** [comment]
@AntoineRelief first version of the context for text widget for editor and summary card`, it would work with any kind of type of field from the given context
The idea is:
- Before sending the data to the parser in both components, check if context data exists, 
 - if it does, we fetch the context fields and selected context resource metadata for those fields and integrate it in the sent fields and fieldsValue to the parser. Also we replace the `{{data.context.property}}` to `{{data.property}}` in the html, this way the parser would work with the additional values for the context without any more logic update
 - If does not, default parser logic applies

I did not add any info alerts on how to add the context, and I did not add the context fields in the autocomplete.
Let me know if it's fine and I'll do that later

KR, Unai

**Comment by AntoineRelief** [comment]
@unai-reliefapp 

that's a good proposition but I'd prefer you go first with the solution I suggested in the ticket, without needing to download any additional data
the reason is your code may have some unwanted impact on the existing configuration, and, because we don't have enough time to correctly check that, I prefer a simple solution

**Comment by unai-reliefapp** [comment]
@AntoineRelief  If i work with `{{context` among any other placeholder, the `updateSettingsContextContent` in the context service triggers and there is no way to access the fields for setting the buttons.
I'll follow this approach but keep a separate step in the parseHtml with a new property named `files`

File: libs/shared/src/lib/components/widgets/editor/editor.component.html [hunk 1]
```diff
@@ -3,7 +3,6 @@
   [ngClass]="{
     'p-[14px]': usePadding
   }"
-  (click)="onClick($event)"
```
Comment (unai-reliefapp) [inline libs/shared/src/lib/components/widgets/editor/editor.component.html:6]: This was causing a double click event trigger, i think this one can be removed(if im not aware of something not clear at first sight) @AntoineRelief

File: libs/shared/src/lib/services/html-parser/html-parser.service.ts [hunk 2]
```diff
@@ -353,30 +353,29 @@ export class HtmlParserService {
         }
       });
 
-      for (const field of fields) {
-        const toReadableObject = (obj: any): any => {
```
Comment (unai-reliefapp) [inline libs/shared/src/lib/services/html-parser/html-parser.service.ts:357]: Moved outside the for loop, it was declared on each track, which does not have sense and was consuming plenty of resources

**Comment by AntoineRelief** [comment]
hey @unai-reliefapp 

do you still have somewhere your commits about fetching the metadata from the resource in order to associate the context fields with resource fields?

we may come to that solution, based on the new requirements we just got

**Comment by unai-reliefapp** [comment]
> hey @unai-reliefapp
> 
> do you still have somewhere your commits about fetching the metadata from the resource in order to associate the context fields with resource fields?
> 
> we may come to that solution, based on the new requirements we just got

You mean the first proposed solution? That one is here: https://github.com/ReliefApplications/ems-frontend/pull/2676/commits/aeb8280d06ba730abac7490b6c0b41d3efa56565

**Comment by AntoineRelief** [comment]
> > hey @unai-reliefapp
> > do you still have somewhere your commits about fetching the metadata from the resource in order to associate the context fields with resource fields?
> > we may come to that solution, based on the new requirements we just got
> 
> You mean the first proposed solution? That one is here: [aeb8280](https://github.com/ReliefApplications/ems-frontend/commit/aeb8280d06ba730abac7490b6c0b41d3efa56565)

thanks @unai-reliefapp !

**Pull request closed without merging**